File: Bridge/Resources/Collections/Dictionary.js
// @source /Collections/Dictionary.js

Bridge.Class.generic('Bridge.KeyValuePair$2', function (TKey, TValue) {
    var $$name = Bridge.Class.genericName('Bridge.KeyValuePair$2', TKey, TValue);

    return Bridge.Class.cache[$$name] || (Bridge.Class.cache[$$name] = Bridge.define($$name, {
        constructor: function (key, value) {
            this.key = key;
            this.value = value;
        },

        toString: function() {
            var s = "[";
            
            if( this.key != null) {
                s += this.key.toString();
            }

            s += ", ";
            if(this.value != null) {
                s += this.value.toString();
            }
            s += "]";
            return s;
        }
    }));
});

Bridge.Class.generic('Bridge.Dictionary$2', function (TKey, TValue) {
    var $$name = Bridge.Class.genericName('Bridge.Dictionary$2', TKey, TValue);

    return Bridge.Class.cache[$$name] || (Bridge.Class.cache[$$name] = Bridge.define($$name, {
        inherits: [Bridge.IDictionary$2(TKey, TValue)],

        constructor: function (obj, comparer) {
            this.comparer = comparer || Bridge.EqualityComparer$1.$default;
            this.clear();

            if (Bridge.is(obj, Bridge.Dictionary$2(TKey, TValue))) {
                var e = Bridge.getEnumerator(obj),
                    c;

                while (e.moveNext()) {
                    c = e.getCurrent();
                    this.add(c.key, c.value);
                }
            } else if (Object.prototype.toString.call(obj) === '[object Object]') {
                var names = Bridge.getPropertyNames(obj),
                    name;

                for (var i = 0; i < names.length; i++) {
                    name = names[i];
                    this.add(name, obj[name]);
                }
            }
        },

        getKeys: function () {
            return new Bridge.DictionaryCollection$1(TKey)(this, true);
        },

        getValues: function () {
            return new Bridge.DictionaryCollection$1(TValue)(this, false);
        },

        clear: function () {
            this.entries = { };
            this.count = 0;
        },

        findEntry: function (key) {
            var hash = this.comparer.getHashCode(key),
                entries,
                i;

            if (Bridge.isDefined(this.entries[hash])) {
                entries = this.entries[hash];

                for (i = 0; i < entries.length; i++) {
                    if (this.comparer.equals(entries[i].key, key)) {
                        return entries[i];
                    }
                }
            }
        },

        containsKey: function (key) {
            return !!this.findEntry(key);
        },

        containsValue: function (value) {
            var e, i;

            for (e in this.entries) {
                if (this.entries.hasOwnProperty(e)) {
                    var entries = this.entries[e];

                    for (i = 0; i < entries.length; i++) {
                        if (this.comparer.equals(entries[i].value, value)) {
                            return true;
                        }
                    }
                }
            }

            return false;
        },

        get: function (key) {
            var entry = this.findEntry(key);

            if (!entry) {
                throw new Bridge.KeyNotFoundException('Key ' + key + ' does not exist.');
            }

            return entry.value;
        },

        getItem: function (key) {
            return this.get(key);
        },

        set: function (key, value, add) {
            var entry = this.findEntry(key),
                hash;

            if (entry) {
                if (add) {
                    throw new Bridge.ArgumentException('Key ' + key + ' already exists.');
                }

                entry.value = value;
                return;
            }

            hash = this.comparer.getHashCode(key);
            entry = new Bridge.KeyValuePair$2(TKey, TValue)(key, value);

            if (this.entries[hash]) {
                this.entries[hash].push(entry);
            } else {
                this.entries[hash] = [entry];
            }

            this.count++;
        },

        setItem: function (key, value, add) {
            this.set(key, value, add);
        },

        add: function (key, value) {
            this.set(key, value, true);
        },

        remove: function (key) {
            var hash = this.comparer.getHashCode(key),
                entries,
                i;

            if (!this.entries[hash]) {
                return false;
            }

            entries = this.entries[hash];

            for (i = 0; i < entries.length; i++) {
                if (this.comparer.equals(entries[i].key, key)) {
                    entries.splice(i, 1);

                    if (entries.length == 0) {
                        delete this.entries[hash];
                    }

                    this.count--;

                    return true;
                }
            }

            return false;
        },

        getCount: function () {
            return this.count;
        },

        getComparer: function () {
            return this.comparer;
        },

        tryGetValue: function (key, value) {
            var entry = this.findEntry(key);

            value.v = entry ? entry.value : Bridge.getDefaultValue(TValue);

            return !!entry;
        },

        getCustomEnumerator: function (fn) {
            var hashes = Bridge.getPropertyNames(this.entries),
                hashIndex = -1,
                keyIndex;

            return new Bridge.CustomEnumerator(function () {
                if (hashIndex < 0 || keyIndex >= (this.entries[hashes[hashIndex]].length - 1)) {
                    keyIndex = -1;
                    hashIndex++;
                }

                if (hashIndex >= hashes.length) {
                    return false;
                }

                keyIndex++;

                return true;
            }, function () {
                return fn(this.entries[hashes[hashIndex]][keyIndex]);
            }, function () {
                hashIndex = -1;
            }, null, this);
        },

        getEnumerator: function () {
            return this.getCustomEnumerator(function (e) {
                 return e;
            });
        }
    }));
});

Bridge.Class.generic('Bridge.DictionaryCollection$1', function (T) {
    var $$name = Bridge.Class.genericName('Bridge.DictionaryCollection$1', T);

    return Bridge.Class.cache[$$name] || (Bridge.Class.cache[$$name] = Bridge.define($$name, {
        inherits: [Bridge.ICollection$1(T)],

        constructor: function (dictionary, keys) {
            this.dictionary = dictionary;
            this.keys = keys;
        },

        getCount: function () {
            return this.dictionary.getCount();
        },

        getEnumerator: function () {
            return this.dictionary.getCustomEnumerator(this.keys ? function (e) {
                return e.key;
            } : function (e) {
                return e.value;
            });
        },

        contains: function (value) {
            return this.keys ? this.dictionary.containsKey(value) : this.dictionary.containsValue(value);
        },

        add: function (v) {
            throw new Bridge.NotSupportedException();
        },

        clear: function () {
            throw new Bridge.NotSupportedException();
        },

        remove: function () {
            throw new Bridge.NotSupportedException();
        }
    }));
});

File: Bridge/Resources.Lua/CoreSystem/Collections/Dictionary.lua
local System = System
local new = System.new
local throw = System.throw
local Collection = System.Collection
local wrap = Collection.wrap
local unWrap = Collection.unWrap
local changeVersion = Collection.changeVersion
local addCount = Collection.addCount
local clearCount = Collection.clearCount
local ArgumentNullException = System.ArgumentNullException
local ArgumentException = System.ArgumentException
local KeyNotFoundException = System.KeyNotFoundException
local EqualityComparer_1 = System.EqualityComparer_1

local Dictionary = {}

local function buildFromCapacity(this, capacity, comparer)
    if comparer ~= nil then
        assert(false)
    end
end

local function buildFromDictionary(this, dictionary, comparer)
    buildFromCapacity(this, 0, comparer)
    if dictionary == nil then
        throw(ArgumentNullException("dictionary"))
    end
    local count = 0
    for k, v in pairs(dictionary) do
        this[k] = v
        count = count + 1
    end
    addCount(this, count)
end

function Dictionary.__ctor__(this, ...) 
    local len = select("#", ...)
    if len == 0 then
        buildFromCapacity(this, 0)
    elseif len == 1 then
        local comparer = ...
        if comparer == nil or type(comparer) == "number" then  
            buildFromCapacity(this, comparer)
        else
            local getHashCode = comparer.getHashCode
            if getHashCode == nil then
                buildFromDictionary(this, comparer)
            else
                buildFromCapacity(this, 0, comparer)
            end
        end
    else 
        local dictionary, comparer = ...
        if type(dictionary) == "number" then 
            buildFromCapacity(this, dictionary, comparer)
        else
            buildFromDictionary(this, dictionary, comparer)
        end
    end
end 

local function checkKey(key)
    if key == nil then
        throw(ArgumentNullException("key"))
    end
end

function Dictionary.add(this, key, value)
    checkKey(key)
    if this[key] then
        throw(ArgumentException("key already exists"))
    end
    this[key] = wrap(value)
    addCount(this, 1)
    changeVersion(this)
end

function Dictionary.clear(this)
    for k, v in pairs(this) do
        this[k] = nil
    end
    clearCount(this)
    changeVersion(this)
end

function Dictionary.containsKey(this, key)
    checkKey(key)
    return this[key] ~= nil 
end

function Dictionary.containsValue(this, value)
    if value == nil then
        for _, v in pairs(this) do
            if unWrap(v) == nil then
                return true
            end
        end    
    else    
        local c = EqualityComparer_1(this.__genericTValue__).getDefault()
        for _, v in pairs(this) do
            if c.equals(value, unWrap(v)) then
                return true
            end
        end
    end
    return false
end

function Dictionary.remove(this, key)
    checkKey(key)
    if this[key] then
        this[key] = nil
        addCount(this, -1)
        changeVersion(this)
        return true
    end
    return false
end

local function getValueDefault(this)
    return this.__genericTValue__.__defaultVal__
end

function Dictionary.tryGetValue(this, key)
    checkKey(key)
    local value = this[key]
    if value == nil then
        return false, getValueDefault(this)
    end
    return true, unWrap(value)
end

function Dictionary.getComparer(this)
    return EqualityComparer_1(this.__genericTKey__).getDefault()
end

Dictionary.getCount = Collection.getCount

function Dictionary.getItem(this, key)
    checkKey(key)
    local value = this[key]
    if value == nil then
        throw(KeyNotFoundException())
    end
    return unWrap(value) 
end

function Dictionary.setItem(this, key, value)
    checkKey(key)
    this[key] = wrap(value)
    changeVersion(this)
end

Dictionary.get = Dictionary.getItem
Dictionary.set = Dictionary.setItem

function Dictionary.getEnumerator(this)
    return Collection.dictionaryEnumerator(this, 0)
end 

local DictionaryCollection = {}
DictionaryCollection.__index = DictionaryCollection

function DictionaryCollection.__ctor__(this, dict, isKey)
    this.dict = dict
    this.isKey = isKey
end

function DictionaryCollection.getCount(this)
    return #this.dict
end

function DictionaryCollection.getEnumerator(this)
    return Collection.DictionaryEnumerator(this.dict, this.isKey and 1 or 2)
end

function DictionaryCollection.contains(this, v)
    if this.isKey then
        return this.dict.containsKey(v)
    end 
    return this.dict.containsValue(v)
end

function Dictionary.getKeys(this)
    return new(DictionaryCollection, this, true)
end

function Dictionary.getValues(this)
    return new(DictionaryCollection, this, false)
end

System.define("System.Dictionary", function(TKey, TValue) 
   local cls = { 
    __inherits__ = { System.IDictionary_2(TKey, TValue) }, 
    __genericTKey__ = TKey,
    __genericTValue__ = TValue,
    __len = Dictionary.getCount
    }
    return cls
end, Dictionary)




